### PR TITLE
Remove websockets dependency & do some refactoring

### PR DESCRIPTION
**Pull request by jack60612**
replaces #6734 

Also a mild clean up

### CODE REVIEW

**Comment by mariano54** [comment]
Has the Unclosed client session error been fixed in RPC server?

**Comment by jack60612** [comment]
> Has the Unclosed client session error been fixed in RPC server?

it has now been fixed

File: chia/rpc/rpc_server.py [hunk 1]
```diff
@@ -259,30 +258,30 @@ async def connection(self, ws):
...
-                        heartbeat=60,
-                        ssl_context=self.ssl_context,
-                        max_msg_size=100 * 1024 * 1024,
-                    ) as ws:
-                        self.websocket = ws
-                        await self.connection(ws)
-                    self.websocket = None
+                self.client_session = aiohttp.ClientSession()
+                self.websocket = await self.client_session.ws_connect(
+                    f"wss://{self_hostname}:{daemon_port}",
```
Comment (emlowe) [inline chia/rpc/rpc_server.py:266]: Wonder why you changed this code as I think the "with" pattern is preferred in general as cleanup is handled for you

File: chia/daemon/client.py [hunk 2]
```diff
@@ -46,7 +61,9 @@ async def _get(self, request: WsRpcMessage) -> WsRpcMessage:
         request_id = request["request_id"]
         self._request_dict[request_id] = asyncio.Event()
         string = dict_to_json_str(request)
-        asyncio.create_task(self.websocket.send(string))
+        if self.websocket is None:
```
Comment (emlowe) [inline chia/daemon/client.py:64]: Won't the `asyncio.create_task(self.websocket.send(string))` throw anyway in this case?

File: chia/daemon/server.py [hunk 3]
```diff
@@ -198,52 +205,48 @@ def cancel_task_safe(self, task: Optional[asyncio.Task]):
...
-                self.log.info(f"ConnectionClosedOk. Closing websocket with {service_name} {e}")
-            elif isinstance(e, WebSocketException):
-                self.log.info(f"Websocket exception. Closing websocket with {service_name} {e} {tb}")
-            else:
-                self.log.error(f"Unexpected exception in websocket: {e} {tb}")
-        finally:
-            self.remove_connection(websocket)
-            await websocket.close()
+    async def incoming_connection(self, request):
+        ws: WebSocketResponse = web.WebSocketResponse(max_msg_size=self.daemon_max_message_size, heartbeat=30)
```
Comment (emlowe) [inline chia/daemon/server.py:215]: Does this `heartbeat=30` override the `heartbeat=60` specified in either "connect_to_daemon" or the `daemon/client.py`

File: chia/daemon/client.py (hunk 2 above)
Comment (mariano54) [reply chia/daemon/client.py:64]: why would it throw?

File: chia/daemon/server.py (hunk 3 above)
Comment (mariano54) [reply chia/daemon/server.py:215]: This is replacing the `ping` task that ws being manually executed every 30 seconds, do it does not change the timeout from the daemon side.
The RPC service side is still at 60. I suppose it's fine to have different ping times for each side of the socket.

File: chia/rpc/rpc_server.py (hunk 1 above)
Comment (mariano54) [reply chia/rpc/rpc_server.py:266]: We were having some issues with `with`, I believe it might be related to saving the object to the class in with, and reusing it later.

Comment (mariano54) [reply chia/rpc/rpc_server.py:266]: This also potentially fixes the annoying error logging when shutting down the wallet.

File: chia/daemon/client.py [hunk 4]
```diff
@@ -18,26 +18,41 @@ def __init__(self, uri: str, ssl_context: Optional[ssl.SSLContext]):
...
-        self.websocket = await websockets.connect(self._uri, max_size=None, ssl=self.ssl_context)
+        try:
+            self.client_session = aiohttp.ClientSession()
+            self.websocket = await self.client_session.ws_connect(
+                self._uri,
+                autoclose=True,
+                autoping=True,
+                heartbeat=60,
+                ssl_context=self.ssl_context,
+                max_msg_size=100 * 1024 * 1024,
```
Comment (mariano54) [inline chia/daemon/client.py:37]: We should probably hardcode this setting like at the top of the file with a constant.

File: chia/daemon/server.py (hunk 3 above)
Comment (mariano54) [reply chia/daemon/server.py:215]: Maybe we should hardcode these constants at the top of the file as well (heartbeat, keepalive timeout etc)

**Review by mariano54 — CHANGES_REQUESTED** [review]
I tried stopping and restarting the wallet, but it won't let me restart because it thinks the wallet is still running. To replicate:

```
chia start wallet
chia stop wallet
chia start wallet
```

**Comment by jack60612** [comment]
> I tried stopping and restarting the wallet, but it won't let me restart because it thinks the wallet is still running. To replicate:
> 
> 
> 
> ```
> 
> chia start wallet
> 
> chia stop wallet
> 
> chia start wallet
> 
> ```

I tried that before pushing, strange

File: chia/rpc/rpc_server.py (hunk 1 above)
Comment (jack60612) [reply chia/rpc/rpc_server.py:266]: Yeah it seems to fix the not closed error, and I think it looks better for this use case

File: chia/daemon/client.py (hunk 2 above)
Comment (emlowe) [reply chia/daemon/client.py:64]: if `self.websocket is none` - then I would expect `asyncio.create_task(self.websocket.send(string))` to throw so one approach is to not bother with the "if" check and just let it throw. Which I assume the previous code does.

File: chia/daemon/server.py (hunk 3 above)
Comment (emlowe) [reply chia/daemon/server.py:215]: Oh right, I didn't notice that ping task - nice to get rid of that.

File: chia/daemon/client.py (hunk 2 above)
Comment (jack60612) [reply chia/daemon/client.py:64]: I'll just do that.

Comment (jack60612) [reply chia/daemon/client.py:64]: correction, one of the linters wants it.

**Comment by jack60612** [comment]
100% working.

**Comment by jack60612** [comment]
All 100% working on my end, should be ready.

File: chia/daemon/server.py [hunk 5]
```diff
@@ -656,18 +634,18 @@ def plot_queue_to_payload(self, plot_queue_item, send_full_log: bool) -> Dict[st
             item["log"] = plot_queue_item.get("log")
         return item
 
-    def prepare_plot_state_message(self, state: PlotEvent, id):
+    def prepare_plot_state_message(self, state: PlotEvent, plot_task_id):
```
Comment (emlowe) [inline chia/daemon/server.py:637]: This seems unrelated - we should remove from this PR?

File: chia/daemon/server.py [hunk 6]
```diff
@@ -745,10 +725,7 @@ def _common_plotting_command_args(self, request: Any, ignoreCount: bool) -> List
         p = request.get("p")  # Pool pubkey
         c = request.get("c")  # Pool contract address
 
-        command_args: List[str] = []
-        command_args.append(f"-n{n}")
-        command_args.append(f"-d{d}")
-        command_args.append(f"-r{r}")
+        command_args: List[str] = [f"-n{n}", f"-d{d}", f"-r{r}"]
```
Comment (emlowe) [inline chia/daemon/server.py:728]: more unrelated changes

File: chia/daemon/server.py (hunk 5 above)
Comment (jack60612) [reply chia/daemon/server.py:637]: Me and mariano did those, having a variable named id can screw up python stuff and can cause errors.

Comment (emlowe) [reply chia/daemon/server.py:637]: Please do it in another PR; it may be a good idea, it's just wholly unrelated to this PR

Comment (jack60612) [reply chia/daemon/server.py:637]: > Please do it in another PR; it may be a good idea, it's just wholly unrelated to this PR

ok, time for pain.

Comment (hoffmang9) [reply chia/daemon/server.py:637]: > ok, time for pain.

Builds character.

Comment (jack60612) [reply chia/daemon/server.py]: done

File: chia/daemon/server.py (hunk 6 above)
Comment (jack60612) [reply chia/daemon/server.py]: fixed

**Review by mariano54 — COMMENTED** [review]
In general I think this is a good change. maybe @emlowe or @Yostra wants to review first before merging. There are not a lot of automated tests of daemon so this requires some manual testing

**Comment by jack60612** [comment]
> It's working for me, but I'm still getting `Unclosed client session` when doing:
> 
> `chia stop wallet`.
> 
> 
> 
> 

I'll check that again, thought I solved that

File: chia/daemon/server.py [hunk 7]
```diff
@@ -1148,22 +1126,18 @@ async def is_running(self, request: Dict[str, Any]) -> Dict[str, Any]:
...
         for k in self.services.keys():
             jobs.append(kill_service(self.root_path, self.services, k))
         if jobs:
             await asyncio.wait(jobs)
         self.services.clear()
+        if self.websocket_runner is not None:
+            await self.websocket_runner.cleanup()
+        self.shutdown_event.set()
 
-        # TODO: fix this hack
```
[the remark this answers is not in the document]
Comment (jack60612) [reply chia/daemon/server.py:1159]: yeah basically, the new WebSocket library does it differently so this old way does not work.

File: chia/daemon/server.py [hunk 8]
```diff
@@ -228,28 +230,27 @@ async def safe_handle(self, websocket: WebSocketServerProtocol, path: str):
...
-        finally:
-            self.remove_connection(websocket)
-            await websocket.close()
+                service_name = "Unknown"
+                if ws in self.remote_address_map:
+                    service_name = self.remote_address_map[ws]
+                if msg.type == WSMsgType.CLOSE:
+                    self.log.info(f"ConnectionClosed. Closing websocket with {service_name}")
+                elif msg.type == WSMsgType.ERROR:
+                    print(f"Websocket exception. Closing websocket with {service_name}. {ws.exception()}")
```
[the remark this answers is not in the document]
Comment (jack60612) [reply chia/daemon/server.py:247]: oops

File: chia/daemon/server.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jack60612) [reply chia/daemon/server.py:1159]: i made a whole new way which fixes the errors we were encountering before.

File: chia/daemon/server.py [hunk 9]
```diff
@@ -184,16 +183,19 @@ def master_close_cb():
...
+        app = web.Application(client_max_size=self.daemon_max_message_size)
+        app.add_routes([web.get("/", self.incoming_connection)])
+        self.websocket_runner = web.AppRunner(app, access_log=None, logger=self.log, keepalive_timeout=300)
+        await self.websocket_runner.setup()
+
+        site = web.TCPSite(
+            self.websocket_runner,
+            host=self.self_hostname,
+            port=self.daemon_port,
+            shutdown_timeout=3,
```
[the remark this answers is not in the document]
Comment (jack60612) [reply chia/daemon/server.py:195]: hmm, looking at the original pr.

[the remark this answers is not in the document]
Comment (jack60612) [reply chia/daemon/server.py:195]: its the amount of time that the server will wait for all connections to stop peacefully during shutdown

**Comment by mariano54** [comment]
can we update / rebase again @jack60612 ?

**Comment by jack60612** [comment]
> can we update / rebase again @jack60612 ?

Done

**Review by mariano54 — APPROVED** [review]
